yt/geometry/oct_visitors.pyx

```diff
@@ -336,3 +336,172 @@ cdef class MortonIndexOcts(OctVisitor):
+        cdef np.int64_t ipos
+        cdef np.float64_t fcoords[3]
+        cdef Oct *neighbour
+        cdef bint local_oct
+        cdef bint other_oct
+        dx = 1.0 / ((1 << self.oref) << self.level)
+        local_oct = True
+
+        # Compute position of neighbouring cell
+        for i in range(3):
+            c = <np.float64_t> (self.pos[i] << self.oref)
+            fcoords[i] = (c + 0.5 + ishift[i]) * dx / self.octree.nn[i]
+            # Assuming periodicity
+            if fcoords[i] < 0:
+                fcoords[i] += 1
+            elif fcoords[i] > 1:
+                fcoords[i] -= 1
+            local_oct &= (0 <= ishift[i] <= 1)
+        other_oct = not local_oct
+
+        # Use octree to find neighbour
+        if other_oct:
+            neighbour = self.octree.get(fcoords, &self.oi, max_level=self.level)
+        else:
+            neighbour = o
+            self.oi.level = self.level
+            for i in range(3):
+                self.oi.ipos[i] = (self.pos[i] << self.oref) + ishift[i]
+
+        # Extra step - compute cell position in neighbouring oct (and store in oi.ipos)
+        if self.oi.level == self.level - 1:
+            for i in range(3):
+                ipos = (((self.pos[i] << self.oref) + ishift[i])) >> 1
+                if (self.oi.ipos[i] << 1) == ipos:
+                    self.oi.ipos[i] = 0
+                else:
+                    self.oi.ipos[i] = 1
+        self.neighbour = neighbour
+
+        # Index of neighbouring cell within its oct
+        for i in range(3):
+            self.neigh_ind[i] = <np.uint8_t>(ishift[i]) % 2
+
+        self.other_oct = other_oct
+        if other_oct:
+            if self.neighbour != NULL:
+                if self.oi.level == self.level - 1:
+                    # Position within neighbouring oct is stored in oi.ipos
+                    for i in range(3):
+                        self.neigh_ind[i] = self.oi.ipos[i]
+                elif self.oi.level != self.level:
+                    print('This should not happen! %s %s' % (self.oi.level, self.level))
+                    self.neighbour = NULL
+
+# Store neighbouring cell index in current cell
+cdef class NeighbourCellIndexVisitor(BaseNeighbourVisitor):
+    @cython.boundscheck(False)
+    @cython.wraparound(False)
+    @cython.initializedcheck(False)
+    cdef void visit(self, Oct* o, np.uint8_t selected):
+        cdef int i, j, k
+        cdef int ishift[3]
+        cdef np.uint8_t neigh_cell_ind
+        cdef np.int64_t neigh_domain_ind
+        if selected == 0: return
+        # Work at oct level
+        if self.last == o.domain_ind: return
+
+        self.last = o.domain_ind
+
+        # Loop over cells in and directly around oct
+        for i in range(-1, 3):
+            ishift[0] = i
+            for j in range(-1, 3):
+                ishift[1] = j
+                for k in range(-1, 3):
+                    ishift[2] = k
+                    self.set_neighbour_info(o, ishift)
+
```

Comment:
This piece of code is very much optimizable. Here are possible routes to achieve much better performance:
1. Work oct by oct, which would reduce the number of neighbor lookup from 4³=64 to 3³=27,
2. Use faster neighbor lookup method(s). For now, all searches are started from the root mesh down to leaf nodes, but we could instead go up the tree from the central oct then down to find all neighbors (see e.g. https://geidav.wordpress.com/2017/12/02/advanced-octrees-4-finding-neighbor-nodes/). 
3. Pre-compute the face-neighbors of *all* octs.

Note that for the last point, algorithms exist that generate the neighbors of all octs in O(1) time (https://link.springer.com/article/10.1007/s13319-015-0060-9) during the octree construction.
Another possible solution would be to keep a unordered hash map of all the octs indexed by their (3-integers) position. With such structure, finding a neighbor takes O(1) time. This could even come as a replacement of the current pointer-based octree structure.

Reply:
I think this should be in a comment.  It obviously doesn't need to be implemented right now.

Reply:
Done.

Reply:
Done!